darkcluster/src/main/java/com/linkedin/darkcluster/util/ExpiringCircularBuffer.java

```diff
@@ -0,0 +1,143 @@
+/*
+   Copyright (c) 2021 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.darkcluster.util;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Queue;
+import org.apache.commons.lang3.NotImplementedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * TODO: javadocs
+ * @param <E>
+ */
+@SuppressWarnings("serial")
+public class ExpiringCircularBuffer<E> extends ArrayList<E> implements Queue<E> {
```

Comment:
IMO, since this is a util class designed primarily for dark cluster use case, it might be better to make this class self-contained so we don't have to implement a lot of unused/unsupported methods.

We probably only need add(), peek(), isFull() and size() for this class.

Reply:
Definitely merits to that. I think it depends on how we end up deduplicating code between SmoothRateLimiter and GuaranteedRateLimiter. Thinking it may be easier to do that if we extend from a common superclass here.

Reply:
The other benefit of not extending ArrayList is that we can have a more granular concurrency control. Currently, we are relying on the parent class implementation of the data structure, so we have to put synchronized key word everywhere.

Reply:
Took your advice here. Next draft will see EvictingCircularBuffer implement a new interface that describes put() and get() methods only, since that's all we use in SmoothRateLimiter as is.